amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.amqp
+
+import java.util.concurrent.atomic.AtomicReference
+
+import com.rabbitmq.client.Connection
+
+import scala.compat.java8.FunctionConverters._
+
+/**
+ * Only for internal implementations
+ */
+sealed trait AmqpConnectionProvider {
+  def settings: AmqpConnectionSettings
+  def get: Connection
+  def release(connection: Connection): Unit
+}
+
+final object DefaultAmqpConnectionProvider {
+
+  /**
+   * Java API
+   */
+  def create(settings: AmqpConnectionSettings): DefaultAmqpConnectionProvider =
+    DefaultAmqpConnectionProvider(settings: AmqpConnectionSettings)
+}
+
+final case class DefaultAmqpConnectionProvider(settings: AmqpConnectionSettings) extends AmqpConnectionProvider {
+  override def get = settings.getConnection
+  override def release(connection: Connection) = if (connection.isOpen) connection.close()
+}
+
+final object ReusableAmqpConnectionProvider {
+
+  /**
+   * Java API
+   */
+  def create(settings: AmqpConnectionSettings): ReusableAmqpConnectionProvider =
+    ReusableAmqpConnectionProvider(settings: AmqpConnectionSettings)
+
+  /**
+   * Java API
+   */
+  def create(settings: AmqpConnectionSettings, automaticRelease: Boolean): ReusableAmqpConnectionProvider =
+    ReusableAmqpConnectionProvider(settings: AmqpConnectionSettings, automaticRelease)
+}
+
+final case class ReusableAmqpConnectionProvider(settings: AmqpConnectionSettings, automaticRelease: Boolean = true)
+    extends AmqpConnectionProvider {
+  var clients = 0
+  private val cachedConnectionRef = new AtomicReference[Option[Connection]](None)
+
+  override def get =
+    cachedConnectionRef
+      .updateAndGet(
```

Comment:
Doesn't this atomic way of updating guarantee that only a thread is running at once?
Similarly to your proposed solution.

Reply:
The problem is that you side effect inside of the function (see the java doc of `updateAndGet` explicitly saying you cannot do that)

Reply:
The clients update could interleave or even not be seen ever between other threads, meaning that the number is pretty much arbitrary. In addition to that the connect call could be called multiple times in case of contention leading to connections being created and leaked.